code/parse/sexp.cpp

```diff
@@ -5541,6 +5541,80 @@ player *get_player_from_ship_node(int node, bool test_respawns = false, int *net
 	return get_player_from_ship_entry(ship_entry, test_respawns, netplayer_index);
 }
 
+/*
+ * Iterates through a subsystem list and applies the given function to each subsystem, handling the "hull" subsystem as well as (optionally) generic subsystems.  The function's prototype should be:
+ * 
+ * void f(bool is_hull, bool is_sim_hull, bool is_shields, ship_subsys *ss);
+ * 
+ * If any of the booleans are true, ss will be nullptr.
+ */
+template <typename Function>
+void process_ship_subsystems(const ship_registry_entry *ship_entry, bool allow_generic_subsystems, int node, bool iterate_nodes, const char *sexp_operator_name, Function f)
```

Comment:
Given the function signature is constant, I'd prefer f to be of type `std::function<void(bool, bool, bool, ship_subsys*)>`.
On second thought, the first three bool parameters appear to be mutually exclusive, so I'd even more prefer for this to be something like an
``enum class SubsystemType : uint8_t {SUBSYSTEM, HULL, SIM_HULL, SHIELDS}``

Reply:
Also, for later:
Given the frequency of the `any_subsystem` check, how about checking that within `process_ship_subsystems` and returning that as a bool? That way you can get around the need to juggle this bool around the lambdas in the invocation.